pip/req/req_set.py

```diff
@@ -396,17 +397,26 @@ def _check_skip_installed(self, req_to_install, finder):
         # Check whether to upgrade/reinstall this req or not.
         req_to_install.check_if_exists()
         if req_to_install.satisfied_by:
-            skip_reason = 'satisfied (use --upgrade to upgrade)'
-            if self.upgrade:
+            upgrade_allowed = False
+
+            # Determine why upgrading this may not allowed.
+            if req_to_install.is_direct:
+                upgrade_allowed = True
+                skip_reason = 'already satisfied'
+            else:
+                skip_reason = 'skipped as not directly required'
```

Comment:
Maybe refactor that out into a function?

---

Checking whether the latest version is installed is done in the same function.

It should be possible to move the "best installed" check to happen before we determine the message of why we're skipping. Having that extra information should be helpful.